presto-main/src/test/java/com/facebook/presto/spiller/TestGenericPartitioningSpiller.java

```diff
@@ -0,0 +1,178 @@
+                spillContextSupplier,
+                memoryContext)) {
+            testSimpleSpiller(spiller, ImmutableSet.of(1, 2)::contains);
+        }
+    }
+
+    private void testSimpleSpiller(PartitioningSpiller spiller, IntPredicate spillPartitionMask)
+            throws ExecutionException, InterruptedException
+    {
+        RowPagesBuilder builder = RowPagesBuilder.rowPagesBuilder(TYPES);
+        builder.addSequencePage(10, SECOND_PARTITION_START, 5, 10, 15);
+        builder.addSequencePage(10, FIRST_PARTITION_START, -5, 0, 5);
+
+        List<Page> firstSpill = builder.build();
+
+        builder = RowPagesBuilder.rowPagesBuilder(TYPES);
+        builder.addSequencePage(10, THIRD_PARTITION_START, 15, 20, 25);
+        builder.addSequencePage(10, FOURTH_PARTITION_START, 25, 30, 35);
+
+        List<Page> secondSpill = builder.build();
+
+        PartitioningSpiller.PartitioningSpillResult result = spiller.partitionAndSpill(firstSpill.get(0), spillPartitionMask);
+        result.getSpillingFuture().get();
+        assertEquals(result.getUnspilledPositions().size(), 0);
+
+        result = spiller.partitionAndSpill(firstSpill.get(1), spillPartitionMask);
+        result.getSpillingFuture().get();
+        assertEquals(result.getUnspilledPositions().size(), 10);
+
+        result = spiller.partitionAndSpill(secondSpill.get(0), spillPartitionMask);
+        result.getSpillingFuture().get();
+        assertEquals(result.getUnspilledPositions().size(), 0);
+
+        result = spiller.partitionAndSpill(secondSpill.get(1), spillPartitionMask);
+        result.getSpillingFuture().get();
+        assertEquals(result.getUnspilledPositions().size(), 10);
+
+        builder = RowPagesBuilder.rowPagesBuilder(TYPES);
+        builder.addSequencePage(10, SECOND_PARTITION_START, 5, 10, 15);
+
+        List<Page> secondPartition = builder.build();
+
+        builder = RowPagesBuilder.rowPagesBuilder(TYPES);
+        builder.addSequencePage(10, THIRD_PARTITION_START, 15, 20, 25);
+
+        List<Page> thirdPartition = builder.build();
+
+        assertSpilledPages(
+                TYPES,
+                spiller,
+                ImmutableList.of(ImmutableList.of(), secondPartition, thirdPartition, ImmutableList.of()));
+    }
+
+    private void assertSpilledPages(
+            List<Type> types,
+            PartitioningSpiller spiller,
+            List<List<Page>> expectedPartitions)
+    {
+        for (int partition = 0; partition < expectedPartitions.size(); partition++) {
+            List<Page> actualSpill = ImmutableList.copyOf(spiller.getSpilledPages(partition));
+            List<Page> expectedSpill = expectedPartitions.get(partition);
+
+            assertEquals(actualSpill.size(), expectedSpill.size());
+            for (int j = 0; j < actualSpill.size(); j++) {
+                assertPageEquals(types, actualSpill.get(j), expectedSpill.get(j));
+            }
+        }
+    }
+
+    private static class FourPartitionsHashGenerator
```

Comment:
I do not like subclassing of concrete `LocalPartitionGenerator` class. 

Maybe instead parametrizing `GenericPartitioningSpiller` with `LocalPartitionGenerator` you should create a specific interface for that
e.g. SpillPartitionStrategy. 

And then have implementation which is based on `LocalPartitionGenerator` for production code use and mock implementation for test.